File: cli/node/src/cli/commands/memory/delete-all.ts
import type { JsonObject } from '../../../backend/json.js';
import type { Backend } from '../../../backend/types.js';
import { printError, printInfo, printSuccess, timedStatus } from '../../../output/branding.js';
import { formatAgentEnvelope, formatJson } from '../../../output/format.js';
import { exit } from '../../../runtime/exit.js';
import { setCurrentCommand } from '../../../runtime/state.js';

/**
 * 校验删除范围和确认条件后批量删除记忆。
 * @param backend - 已装配的后端连接器。
 * @param opts - 本次操作选项。
 * @param opts.force - 跳过已有配置或删除操作的交互确认。
 * @param opts.dryRun - 仅展示删除目标，不执行删除。
 * @param opts.all - 请求项目范围的删除操作。
 * @param opts.userId - 用户 ID。
 * @param opts.agentId - 代理 ID。
 * @param opts.appId - 应用 ID。
 * @param opts.runId - 会话 ID。
 * @param opts.output - 终端输出格式。
 */
export async function cmdDeleteAll(
  backend: Backend,
  opts: {
    force: boolean;
    dryRun?: boolean;
    all?: boolean;
    userId?: string;
    agentId?: string;
    appId?: string;
    runId?: string;
    output: string;
  }
): Promise<void> {
  setCurrentCommand('delete-all');
  if (opts.all && opts.dryRun) {
    printError('Project-wide --dry-run is not supported; no memories were deleted.');
    exit(1);
  }
  const { isAgentMode } = await import('../../../runtime/state.js');
  if (isAgentMode() && !opts.force) {
    printError('Destructive operation requires --force in agent mode.');
    exit(1);
  }
  if (opts.all) {
    // 项目级删除使用通配实体标识。
    // 历史行为：项目级删除不支持预先计数；相关语义必须显式验收。

    if (!opts.force) {
      const readline = await import('node:readline');
      const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
      });
      const answer = await new Promise<string>((resolve) => {
        rl.question(
          '\n  \u26a0  Delete ALL memories across the ENTIRE project? This cannot be undone. [y/N] ',
          resolve
        );
      });
      rl.close();
      if (answer.toLowerCase() !== 'y') {
        printInfo('Cancelled.');
        exit(0);
      }
    }

    const start = performance.now();
    let result: JsonObject;
    try {
      result = await timedStatus('Deleting all memories project-wide...', async () => {
        return backend.delete(undefined, {
          all: true,
          userId: '*',
          agentId: '*',
          appId: '*',
          runId: '*',
        });
      });
    } catch (e) {
      printError(e instanceof Error ? e.message : String(e));
      exit(1);
    }
    const elapsed = (performance.now() - start) / 1000;

    if (opts.output === 'agent') {
      formatAgentEnvelope({
        command: 'delete-all',
        data: result,
        durationMs: Math.round(elapsed * 1000),
      });
    } else if (opts.output === 'json') {
      formatJson(result);
    } else if (opts.output !== 'quiet') {
      if (result.message) {
        printInfo('Deletion started. Memories will be removed in the background.');
      } else {
        printSuccess(`All project memories deleted (${elapsed.toFixed(2)}s)`);
      }
    }
    return;
  }

  if (opts.dryRun) {
    let memories: JsonObject[];
    try {
      memories = await backend.listMemories({
        userId: opts.userId,
        agentId: opts.agentId,
        appId: opts.appId,
        runId: opts.runId,
      });
    } catch (e) {
      printError(e instanceof Error ? e.message : String(e));
      exit(1);
    }
    printInfo(`Would delete ${memories.length} memories.`);
    printInfo('No changes made.');
    return;
  }

  if (!opts.force) {
    const scopeParts: string[] = [];
    if (opts.userId) scopeParts.push(`user=${opts.userId}`);
    if (opts.agentId) scopeParts.push(`agent=${opts.agentId}`);
    if (opts.appId) scopeParts.push(`app=${opts.appId}`);
    if (opts.runId) scopeParts.push(`run=${opts.runId}`);
    const scope = scopeParts.length > 0 ? scopeParts.join(', ') : 'ALL entities';

    const readline = await import('node:readline');
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
    });
    const answer = await new Promise<string>((resolve) => {
      rl.question(
        `\n  \u26a0  Delete ALL memories for ${scope}? This cannot be undone. [y/N] `,
        resolve
      );
    });
    rl.close();
    if (answer.toLowerCase() !== 'y') {
      printInfo('Cancelled.');
      exit(0);
    }
  }

  const start = performance.now();
  let result: JsonObject;
  try {
    result = await timedStatus('Deleting all memories...', async () => {
      return backend.delete(undefined, {
        all: true,
        userId: opts.userId,
        agentId: opts.agentId,
        appId: opts.appId,
        runId: opts.runId,
      });
    });
  } catch (e) {
    printError(e instanceof Error ? e.message : String(e));
    exit(1);
  }
  const elapsed = (performance.now() - start) / 1000;

  if (opts.output === 'agent') {
    formatAgentEnvelope({
      command: 'delete-all',
      data: result,
      durationMs: Math.round(elapsed * 1000),
    });
  } else if (opts.output === 'json') {
    formatJson(result);
  } else if (opts.output !== 'quiet') {
    if (result.message) {
      printInfo('Deletion started. Memories will be removed in the background.');
    } else {
      printSuccess(`All matching memories deleted (${elapsed.toFixed(2)}s)`);
    }
  }
}

File: cli/python/src/memgo_cli/cli/commands/memory/delete_all.py
"""记忆 delete_all 用例。"""

from __future__ import annotations

import time as _time

import typer
from rich.console import Console

from memgo_cli.backend.types import Backend
from memgo_cli.output.branding import (
    print_error,
    print_info,
    print_success,
    timed_status,
)
from memgo_cli.output.format import (
    format_agent_envelope,
    format_json,
)

console = Console()
err_console = Console(stderr=True)


def cmd_delete_all(
    backend: Backend,
    *,
    force: bool,
    dry_run: bool,
    all_: bool,
    user_id: str | None,
    agent_id: str | None,
    app_id: str | None,
    run_id: str | None,
    output: str,
) -> None:
    """删除匹配范围的全部记忆，支持预览和确认。"""
    from memgo_cli.runtime.state import is_agent_mode, set_current_command

    set_current_command("delete-all")
    if is_agent_mode():
        output = "agent"
        if not force:
            print_error(
                err_console, "Destructive operation requires --force in agent mode.", hint=None
            )
            raise typer.Exit(1)
    if all_:
        # 使用通配符实体 ID 执行项目范围删除
        # 平台没有项目范围预览接口，明确拒绝，绝不把预览执行为删除。
        if dry_run:
            print_error(
                err_console,
                "Project-wide --dry-run is not supported; no memories were deleted.",
                hint=None,
            )
            raise typer.Exit(1)

        if not force:
            confirm = typer.confirm(
                "\n  ⚠  Delete ALL memories across the ENTIRE project? This cannot be undone."
            )
            if not confirm:
                print_info(console, "Cancelled.")
                raise typer.Exit(0)

        _start = _time.perf_counter()
        with timed_status(err_console, "Deleting all memories project-wide...") as _ts:
            try:
                result = backend.delete(
                    all=True,
                    user_id="*",
                    agent_id="*",
                    app_id="*",
                    run_id="*",
                    memory_id=None,
                    delete_linked=False,
                )
            except Exception as e:
                print_error(err_console, str(e), hint=None)
                raise typer.Exit(1) from None
        _elapsed = _time.perf_counter() - _start

        if output == "agent":
            format_agent_envelope(
                console,
                command="delete-all",
                data={"deleted": True, "scope": "project"},
                duration_ms=int(_elapsed * 1000),
                scope=None,
                count=None,
            )
        elif output == "json":
            format_json(console, result)
        elif output != "quiet":
            if isinstance(result, dict) and "message" in result:
                print_info(console, "Deletion started. Memories will be removed in the background.")
            else:
                print_success(console, f"All project memories deleted ({_elapsed:.2f}s)")
        return

    if dry_run:
        # 列出匹配记忆并显示数量
        try:
            results = backend.list_memories(
                user_id=user_id,
                agent_id=agent_id,
                app_id=app_id,
                run_id=run_id,
                page=1,
                page_size=100,
                category=None,
                after=None,
                before=None,
                show_expired=False,
                latest_only=False,
            )
        except Exception as e:
            print_error(err_console, str(e), hint=None)
            raise typer.Exit(1) from None
        count = len(results)
        print_info(console, f"Would delete {count} memor{'y' if count == 1 else 'ies'}.")
        print_info(console, "No changes made (dry run).")
        return

    if not force:
        scope_parts = []
        if user_id:
            scope_parts.append(f"user={user_id}")
        if agent_id:
            scope_parts.append(f"agent={agent_id}")
        if app_id:
            scope_parts.append(f"app={app_id}")
        if run_id:
            scope_parts.append(f"run={run_id}")
        scope = ", ".join(scope_parts) if scope_parts else "ALL entities"

        confirm = typer.confirm(f"\n  ⚠  Delete ALL memories for {scope}? This cannot be undone.")
        if not confirm:
            print_info(console, "Cancelled.")
            raise typer.Exit(0)

    _start = _time.perf_counter()
    with timed_status(err_console, "Deleting all memories...") as _ts:
        try:
            result = backend.delete(
                all=True,
                user_id=user_id,
                agent_id=agent_id,
                app_id=app_id,
                run_id=run_id,
                memory_id=None,
                delete_linked=False,
            )
        except Exception as e:
            print_error(err_console, str(e), hint=None)
            raise typer.Exit(1) from None
    _elapsed = _time.perf_counter() - _start

    scope_ids = {
        k: v
        for k, v in {
            "user_id": user_id,
            "agent_id": agent_id,
            "app_id": app_id,
            "run_id": run_id,
        }.items()
        if v
    }
    if output == "agent":
        format_agent_envelope(
            console,
            command="delete-all",
            data={"deleted": True},
            scope=scope_ids or None,
            duration_ms=int(_elapsed * 1000),
            count=None,
        )
    elif output == "json":
        format_json(console, result)
    elif output != "quiet":
        if isinstance(result, dict) and "message" in result:
            print_info(console, "Deletion started. Memories will be removed in the background.")
        else:
            print_success(console, f"All matching memories deleted ({_elapsed:.2f}s)")
